### ISSUE TITLE
Problem detecting code changes in TS project

### ISSUE THREAD

**Issue by rarkins** [wallabyjs/public#2866]
### Issue description or question

I am running Wallaby.js in this open source project https://github.com/renovatebot/renovate using a paid personal license. The project uses TS + Jest.

First of all, it seemed to autodetect the test files - so far so good.

There's four failures, three of which are somewhat predictable. 

![image](https://user-images.githubusercontent.com/6311784/144670413-67e6a63b-1770-4561-adb6-678888585d3d.png)

We pass `TZ=UTC` to our `yarn jest` command to get those tests to pass. I tried setting up a `wallaby.js` with custom env and re-running the test, but they still failed. Maybe I needed to restart Wallaby.js?

I then rewrote the code so that the custom TZ variable isn't needed for tests to pass. `yarn jest artifactory` now passes without the need for the variable, but unfortunately Wallaby (a) doesn't autodetect and re-run the tests, and (b) doesn't pass them even after a manual change.

Is there something about TS which would mean that the changes aren't getting detected and picked up?

### Wallaby diagnostics report

The report is very long and hangs the tab (in Chrome) when I paste it in so I put it in a gist instead: https://gist.github.com/rarkins/4b980bb0e65c3fd64f6e29a68b9b0d56

**Comment by smcenlly** [comment]
Wallaby runs your code the same as if you were using `npx jest` from the command line. I tried to clone your repo and `npx jest` from the project root failed for me (I'm not sure if this is intentional).

> Test Suites: 262 failed, 116 passed, 378 total
> Tests:       407 failed, 2920 passed, 3327 total
> Snapshots:   1 failed, 217 passed, 218 total

From your diagnostics report, it looks like Wallaby is watching a few other folders that it shouldn't (e.g. `.cache`, and `coverage`). Could you please add these folders to the `watchPathIgnorePatterns` section of jest configuration (you should be able to confirm it's working by looking at your diagnostics report). I think what may be happening (and why the changes don't seem to be detected) is that Wallaby starts running your tests again, then detects that these files have been changed, cancels the test runs, and starts again (so appears to be doing thing).

**Comment by rarkins** [comment]
Thanks for the quick feedback. Just to confirm, do you mean I should aim for the following to work?
- clone
- yarn install
- npx jest

I wanted to clarify if/how the node_modules are meant to be populated, or how that factors into it. Also our current `yarn jest` is a bit more complicated than that but I'll see what can be done.

**Comment by rarkins** [comment]
When I run `npx jest` against this repo - with a fully populated `node_modules` - then I get the same four failures as Wallaby presents in its UI. So this seems to provide me with a good approach for solving the test failure aspect, and I'll address the watch afterwards.

**Comment by rarkins** [comment]
I now have added `.cache` and `coverage` to the ignore patterns - or at least tried: https://github.com/renovatebot/renovate/blob/fc3461d5edbc605cea501795dabc1987a17960bd/jest.config.ts#L37

However the wallaby.js diagnostics log is still filled with `.cache` references, e.g.

```
    '2021-12-04T12:41:01.535Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/b6/requiredstatuschecksmigration_b65db03d38fca754dffbeebf8d06d1a4\n',
```

I tried adding these to both `testPathIgnorePatterns` as well as `watchPathIgnorePatterns`.

I'm really confused by this, because I wouldn't expect to need to explicitly ignore these:
- `.cache` starts with a dot, which I thought was automatically ignored
- It's configured in the same config file as jest's `cacheDirectory`.. which you'd think should get added to jest watch ignore automatically?

Snippet from latest diagnostics after a restart:
```
    '2021-12-04T12:50:47.378Z workers [worker #0, session #ao6nk] Running tests in sandbox\n',
    '2021-12-04T12:50:47.382Z workers [worker #1, session #c1jwh] Running tests in sandbox\n',
    '2021-12-04T12:50:47.385Z workers [worker #2, session #yjglz] Running tests in sandbox\n',
    '2021-12-04T12:50:47.389Z workers [worker #3, session #bcpr0] Running tests in sandbox\n',
    '2021-12-04T12:50:47.393Z workers [worker #4, session #mapid] Running tests in sandbox\n',
    '2021-12-04T12:50:47.396Z workers [worker #5, session #ly6kw] Running tests in sandbox\n',
    '2021-12-04T12:50:47.400Z workers [worker #6, session #pzt9g] Running tests in sandbox\n',
    '2021-12-04T12:50:47.403Z workers [worker #7, session #vbr5v] Running tests in sandbox\n',
    '2021-12-04T12:50:47.406Z workers [worker #8, session #6xe6t] Running tests in sandbox\n',
    '2021-12-04T12:50:47.408Z workers [worker #9, session #v4sbq] Running tests in sandbox\n',
    '2021-12-04T12:51:02.649Z fs No metadata for added file found: .cache/jest/haste-map-193e715aefacde94321c173cc4efaf8d-b0c5951cbfdbcd6eda9e09c684e1b169-0451578f793c8d987498da9c18e8ece4\n',
    '2021-12-04T12:51:02.749Z fs File changed: .cache/jest/haste-map-193e715aefacde94321c173cc4efaf8d-b0c5951cbfdbcd6eda9e09c684e1b169-0451578f793c8d987498da9c18e8ece4\n',
    '2021-12-04T12:51:02.750Z fs No metadata for changed file found: .cache/jest/haste-map-193e715aefacde94321c173cc4efaf8d-b0c5951cbfdbcd6eda9e09c684e1b169-0451578f793c8d987498da9c18e8ece4\n',
    '2021-12-04T12:51:04.014Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/ba/newlinesnapshotserializer_ba5c7a25b58496a38432288c112ad43a.map\n',
    '2021-12-04T12:51:04.014Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/ba/newlinesnapshotserializer_ba5c7a25b58496a38432288c112ad43a\n',
    '2021-12-04T12:51:04.014Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/54/newlinesnapshotserializer_54f0ccdf39551119ad3b01dddf20a2ca.map\n',
    '2021-12-04T12:51:04.014Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/54/newlinesnapshotserializer_54f0ccdf39551119ad3b01dddf20a2ca\n',
    '2021-12-04T12:51:04.114Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/15/newlinesnapshotserializer_15b3a183adc07aeceff62904333ab66b.map\n',
    '2021-12-04T12:51:04.114Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/15/newlinesnapshotserializer_15b3a183adc07aeceff62904333ab66b\n',
    '2021-12-04T12:51:04.215Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/4d/newlinesnapshotserializer_4dc074ed859f7dfe58bb2f23571338ab.map\n',
    '2021-12-04T12:51:04.215Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/4d/newlinesnapshotserializer_4dc074ed859f7dfe58bb2f23571338ab\n',
    '2021-12-04T12:51:04.215Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/8c/newlinesnapshotserializer_8cd02ba329103e23d588cb6851d6e636.map\n',
    '2021-12-04T12:51:04.215Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/8c/newlinesnapshotserializer_8cd02ba329103e23d588cb6851d6e636\n',
    '2021-12-04T12:51:04.215Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/28/newlinesnapshotserializer_28670a4564f3dcb58343a41ce7636978.map\n',
    '2021-12-04T12:51:04.216Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/28/newlinesnapshotserializer_28670a4564f3dcb58343a41ce7636978\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/42/newlinesnapshotserializer_42326bd99631cce1b70ba7ad60608515.map\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/42/newlinesnapshotserializer_42326bd99631cce1b70ba7ad60608515\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/cf/newlinesnapshotserializer_cf4e81e1ca6582fee00efa85e6d2f999.map\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/cf/newlinesnapshotserializer_cf4e81e1ca6582fee00efa85e6d2f999\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/7e/newlinesnapshotserializer_7e320ab4ae74d94aca55d96e10cf74db.map\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/8b/newlinesnapshotserializer_8bb876e5399269fd0895c00d70d90748.map\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/7e/newlinesnapshotserializer_7e320ab4ae74d94aca55d96e10cf74db\n',
    '2021-12-04T12:51:04.317Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-5bcdb64e20894c4318a9b7ff46ffc22d/8b/newlinesnapshotserializer_8bb876e5399269fd0895c00d70d90748\n',
```

I'm finished experimenting so now waiting on your advice on next steps.

**Comment by smcenlly** [comment]
We have updated Wallaby core (`v1.0.1185`) so that you should no longer need to explicitly specify `cache` or `coverage` folders. This project explicitly sets the `cache` to be within the project directory which we had not encountered before.

We have taken a look at the open source project and Wallaby appears to be working correctly but the tests are very slow. Outside of Wallaby, on my computer, the tests take `4 minutes 38 seconds`. Using Wallaby, the initial test run takes `4 minutes 51 seconds` (we would expect similar results).

If you were to wait after making some changes (may be up to the initial test run time) then I think you will find Wallaby is working (albeit, it is too slow). The problem is actually not a Wallaby issue but is due to the structure of the project and its tests.

When you change a **single test** in the project, Wallaby will pass/fail a single test in about 1/2 a second. When changing application code in the project, because of the file dependency graph, Wallaby determines that it needs to execute many tests within your project due to which files are imported from tests and how the dependencies relate. For example, changing `lib/platform/azure/azure-got-wrapper.ts` (which seems like a fairly leaf-level dependency) results in 101 tests being executed across 11 separate test files.

You should see similar behavior when running the tests from the command-line. If for example you are changing some code that takes 2-3 minutes to run all tests then Wallaby will appear to be not doing anything for 2-3 minutes while it's running your tests. As you are making code changes, Wallaby will also cancel your current test-run and start to re-execute all of the tests again with the updated code-changes.

We generated a dependency graph for this project (it was too large to share - 4MB SVG, 15MB PNG) which show that many tests rely on high-level dependencies which means that many tests need to be re-run on seemingly isolated changes. Ideally, you would optimise the project to run the tests faster and minimise test files/test runs relying on top-level dependencies in your dependency graph, but for this project, I understand there's probably not much you can do.

There are a few things that you may like to try:

1. The `ignoreFileLoadingDependencyTracking ` setting (see [our docs](https://wallabyjs.com/docs/config/overview.html#ignoring-file-loading-chains-for-files-dependency-tracking-purposes)), which will prevent Wallaby from tracking dependencies between files. This may mean that some tests don't get executed on changes when they should.

2. [Running selected tests](https://wallabyjs.com/docs/intro/selected-tests.html#exclusive-test-run), in particular the **Exclusive Test Run** feature.

**Comment by rarkins** [comment]
Meta: please let me know if you prefer topics like the below to be continued here even though the issue is closed, or in a new issue, or even in multiple issues (one per topic).

Thanks for the quick fix. I'm ok with the startup time being long due to the long test time. My decision to buy Wallabyjs was largely *because* the full set of unit tests take so long to run, that I'd like to just run the necessary subset locally before pushing to CI.

FYI I am the maintainer of this project so definitely can put some effort into test refactoring if it will increase productivity of all involved.

Questions:

1. Re: `ignoreFileLoadingDependencyTracking`, we already have a linting rule which means that all imports must be used. In such a case, this setting likely doesn't change anything, right?

2. Is the Exclusive Test Run feature meant as an alternative to the 4 minutes+ default which happens when Wallaby starts?

3. Is there any way for us to debug our dependency graph the way you did, to perhaps find some low hanging fruit? e.g. so I could start with your example `lib/platform/azure/azure-got-wrapper.ts` and find the 11 files / 101 tests without tracing that manually. We don't want to stub/mock every single file, or else our unit testing becomes less effective, but I'm sure there's some inefficient dependency graph branches we could improve.

**Comment by smcenlly** [comment]
> Meta: please let me know if you prefer topics like the below to be continued here even though the issue is closed, or in a new issue, or even in multiple issues (one per topic).

Happy to reply in closed issue if that works for you.

> 1. Re: ignoreFileLoadingDependencyTracking, we already have a linting rule which means that all imports must be used. In such a case, this setting likely doesn't change anything, right?

Yes - I wouldn't expect it to do much to help in this case, but may be worth trying.

> 2. Is the Exclusive Test Run feature meant as an alternative to the 4 minutes+ default which happens when Wallaby starts?

Yes. It is probably ideal for projects like the one that you shared where there are many tests and they take a while to run. The idea is that you know that you're working on a specific area/feature/etc. and you know you only expect to impact a few sets of tests (or perhaps just one test file). In this case, start the exclusive test run on just one test file, or on one directory and Wallaby will only run those tests.

For example, starting an exclusive test run on `lib/platform/azure/azure-got-wrapper.spec.ts` starts Wallaby for just those tests in 3 seconds on my machine. After that, a change to `lib/platform/azure/azure-got-wrapper.ts` re-executes tests run 582ms instead of ~9 seconds.

> 3. Is there any way for us to debug our dependency graph the way you did, to perhaps find some low hanging fruit? e.g. so I could start with your example lib/platform/azure/azure-got-wrapper.ts and find the 11 files / 101 tests without tracing that manually. We don't want to stub/mock every single file, or else our unit testing becomes less effective, but I'm sure there's some inefficient dependency graph branches we could improve.

We used [`ts_dependency_graph`](https://github.com/PSeitz/ts-dependency-graph) to inspect the dependency chains. You may also find [juanallo.vscode-dependency-cruiser](https://marketplace.visualstudio.com/items?itemName=juanallo.vscode-dependency-cruiser) useful if you are using VS Code (although it's not interactive).

**Comment by rarkins** [comment]
Thank you. I will explore the dependency trees and see if we can optimize anything, especially at times when I see tests churning.

Additionally, I now have `npx jest` passing 100% but Wallaby has 1 failure I can't work out:

```
‌1 failing test, 6834 passing, 12 skipped  ​Launch Coverage & Test Explorer​ | ​Search Tests​
  
  ​​​​workers/repository/dependency-dashboard ensureDependencyDashboard() rechecks branches​​​​ ​​​[5 ms]​​​​​ Debug test​ |​ Open test story​
     
    TypeError: Cannot read properties of undefined (reading 'match') 
      ​​​​​at parseDashboardIssue ​​​​​​lib/workers/repository/dependency-dashboard.ts:17:2​
      ​​​​​at Object.ensureDependencyDashboard ​​​​​​lib/workers/repository/dependency-dashboard.ts:332:6​
      ​​​​​at Object.<anonymous> ​​​​​​lib/workers/repository/dependency-dashboard.spec.ts:524:6
```

I really suspect it's a bug, but please let me know if there's any other ways to troubleshoot it.

**Comment by smcenlly** [comment]
> I really suspect it's a bug, but please let me know if there's any other ways to troubleshoot it.

This definitely was a bug, thanks for letting us know about it. Fortunately because your project is open-source we could quickly find and fix the problem. The issue was actually related to [this issue](https://github.com/estools/escodegen/commit/f6df59fc068600803d3a945406de4620b044883a) - Wallaby uses a fork of this project.

The issue is fixed in Wallaby core v1.0.1186.

**Comment by rarkins** [comment]
Thanks, I now have all tests passing.

FYI I'm still seeing a lot of cache noise in the diagnostics, e.g.

```
    '2021-12-08T08:29:33.961Z fs No metadata for added file found: .cache/jest/jest-transform-cache-193e715aefacde94321c173cc4efaf8d-0a7b84a12d1de59a7d5dc1647511900f/6c/resultspec_6c1e4cd3b089e1b8e3040284fa21252a\n',
```

Is that to be expected? I thought you'd want to 100% ignore that cache.

Something else which would be nice is if the diagnostics could print info related to the dependency tree. e.g. if I modify one file and that triggers 11 to be re-run, I would be great if that relationship was logged. That way if I find the Wallaby.js performance unsatisfactory I could follow a process like:
- Observe an excessive re-run duration
- Open Wallabyjs diagnostics and find the list of files to understand why
- Explore the tree manually or through other tools and work out if I should be refactoring to reduce all the links

**Comment by smcenlly** [comment]
> Is that to be expected? I thought you'd want to 100% ignore that cache.

This is expected. I am pretty sure that you shouldn't see `changed` or `deleted` as you saw before.

Generally speaking, we don't expect others to be reading our diagnostics report so unless you're having issues, you can safely ignore the content of your diagnostics report going forward.

> Something else which would be nice is if the diagnostics could print info related to the dependency tree. e.g. if I modify one file and that triggers 11 to be re-run, I would be great if that relationship was logged.

We have spoken about these features for some time, which would also allow us to provide visibility into what tests are running along with an estimate as to how much longer your tests have to run. We do have the information in our internal model(s) but implementing these features is not a quick/simple task. It is definitely something that we want to do though, ideally in the next 6-12 months.
